Make wxDC and wxGraphicsContext DPI aware

MaartenBent opened the pull request:
Add and use `FromDIP` and `ToDIP`, and make returned values from `GetDPI`/`GetPPI` consistent with `wxWindow`.

I am testing this mainly with the drawing sample (polygon page) and printing sample which have been made DPI aware. I am checking if all renderers, saved files, print to pdf, look the same compared to default 96dpi/100%.

This includes:
* msw
  * wxDC, wxGCDC with gdi+, d2d, and cairo
  * with and without wxBufferedPaintDC
  * saving as bmp, svg, ps
  * at different dpi
    * static: main monitor with dpi other dan 96/100%
    * dynamic: move from one dpi to another
  * different dpi in manifest: no, system, per-monitor

* osx
  * wxDC, wxGCDC with cg and cairo
  * with and without wxBufferedPaintDC
  * saving as bmp
  * at 144dpi/retina and with 'Open in Low Resolution' option
 
* gtk3
  * wxDC, wxGCDC with cairo
  * with and without wxBufferedPaintDC
  * saving as bmp
  * with scale 100% and 200%

---
TODO
- [x] Check gtk3
- [x] Update interface files (remove `GetDPIScaleFactor`, add `FromDIP`, `ToDIP`)

MaartenBent committed:
CMake: Enable building with Cairo renderer on MSW without GTK

MaartenBent committed:
CMake: Fix warning when copying inplace config

Quote all parameters to ensure paths are handled properly.

MaartenBent committed:
Recreate DC buffer when content scale factor changes

MaartenBent committed:
Improve DPI support in wxMSW wxCairoRenderer

Drawing sample looks correct on Windows and Linux GTK3 at 200% scaling,
and correct DPI is returned.

MaartenBent committed:
Remove GetDPIScaleFactor from wxDC and wxGraphicsContext

This reverts most of the changes from ee2b02614e (Add GetDPIScaleFactor to wxDC
and wxGraphicsContext, 2022-04-16).

This is not supposed to be used to scale pixels, FromDIP will be added instead.

Temporary use scale=1 in the drawing sample until FromDIP is added.

MaartenBent committed:
Add FromDIP to wxDC and wxGraphicsContext

Using the same implementation as wxWindow has.

MaartenBent committed:
Make the polygon page in the drawing sample DPI aware

Remove unneeded wxGCDC constructor argument.

Make showing a clipping region, textured background, and changing
origin DPI aware.

MaartenBent committed:
Remove trailing whitespace in macOS graphics code

MaartenBent committed:
Override GetDPI in wxMacCoreGraphicsContext

Let it return the DPI of the associated window, or (just like wxWindowMac::GetDPI)
return the DPI derived from the content scale factor.

MaartenBent committed:
Check if GetDPI parameters are not NULL

And combine the same code for assigning the variables.

MaartenBent committed:
Override FromDIP in wxSVGFileDC and wxPostScriptDC

These are DPI independent.

MaartenBent committed:
Improve print preview and printing with high DPI

All print related scaling is based on 96DPI, so always let the printing contexts return this,
and also adjust the font to this DPI.

MaartenBent committed:
Make the printing sample DPI aware

MaartenBent committed:
Only scale MSW wxMemoryDC without window with content scale factor

MaartenBent committed:
Let wxGCDC inherit wxWindow from wxGraphicsContext

MaartenBent committed:
Don't override wxUSE_DPI_AWARE_MANIFEST in resource file

MaartenBent committed:
Fix saving bitmap with cairo renderer on macOS

MaartenBent committed:
Add ToDIP to wxDC and wxGraphicsContext

MaartenBent committed:
Fix bitmap size when saving image in drawing sample

wxBitmap::CreateWithDIPSize expects an DPI independent size.
And when saving as SVG, it also expects a DPI independent size.

Also show DPI independent coordinates in the status bar.

vadz commented on src/msw/dcmemory.cpp:
```diff
@@ -143,9 +143,12 @@ void wxMemoryDCImpl::DoSelect( const wxBitmap& bitmap )
+    if ( GetWindow() )
+        return GetWindow()->GetDPI();
+    else
+        return wxDisplay::GetStdPPI() * GetContentScaleFactor();
```
Sorry, I don't understand this:

1. How can a `wxMemoryDC` have an associated window?
2. Why do we need multiply only the display DPI but not the window one by the scale factor?

MaartenBent replied:
I added this for a workaround I mentioned before (using `memdc.GetImpl()->SetWindow( )`). And that window already has the correct DPI so no scaling would be needed.

It is probably better to remove it, so I did.

vadz commented on src/msw/dcprint.cpp:
```diff
@@ -230,6 +231,13 @@ wxRect wxPrinterDCImpl::GetPaperRect() const
+{
+    wxFont scaledFont = font;
+    if ( scaledFont.IsOk() )
+        scaledFont.WXAdjustToPPI(wxDisplay::GetStdPPI());
```
I'm not sure why is this needed neither...

MaartenBent replied:
Based on testing I found that `wxPrinterDCImpl` and `wxEnhMetaFileDCImpl` below are based on 96DPI. So the font has to be adjusted to this DPI as well.
Otherwise the font size in the print preview and print (to pdf) are too big.

vadz commented on src/msw/enhmeta.cpp:
```diff
@@ -223,6 +224,24 @@ class wxEnhMetaFileDCImpl : public wxMSWDCImpl
+    {
+        wxFont scaledFont = font;
+        if (scaledFont.IsOk())
+            scaledFont.WXAdjustToPPI(wxDisplay::GetStdPPI());
```
Same question, why do we need to do this?

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -1124,15 +1130,22 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
 {
     InitColour(col);
 
+#ifndef __WXMAC__
```
I'd prefer to test for `__WXMAC__` rather than its negation, it's more clear and more consistent with the existing tests below.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -1124,15 +1130,22 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
+        : double(font.GetFractionalPointSize() * dpi.y / 72);
+#else
+    wxUnusedVar(dpi);
     m_size = font.GetPointSize();
```
Why not `GetFractionalPointSize()`?

MaartenBent replied:
Yes, that can be used here.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -1124,15 +1130,22 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
     InitColour(col);
 
+#ifndef __WXMAC__
+    m_size = !dpi.y
+        ? double(font.GetPixelSize().GetHeight())
+        : double(font.GetFractionalPointSize() * dpi.y / 72);
```
Could you please explain why do we do this and why do we do a different thing under Mac?

This must be hiding some more fundamental incompatibility/difference between the platforms, but I don't know which one.

MaartenBent replied:
I think there are two differences here between macOS and Windows.

The first one is that macOS is based on 72dpi and Windows on 96dpi. Fonts are in 72ppi, so on Windows they have to be corrected to the DPI. On macOS no correction is needed.

Second, on Windows the font has to be adjusted to the screen DPI (96 at 100%, 192 at 200%) to get the correct font size. On macOS this is not needed (content scale stuff?) so you basically end up with `font.GetFractionalPointSize() * 72 / 72)`.

vadz replied:
I'd prefer to use `wxDisplay::GetStdPPIValue() / 72`, this should be optimized away under Mac and would result in more clear code IMO.

As for "DPI correction", I'd indeed expect it to be taken care of by `GetContentScaleFactor()`, but its value is the same for GTK and macOS, while this code is Mac-specific. I.e. it's not so much the difference between MSW and Mac that I don't understand, but the difference between Mac and GTK.

MaartenBent replied:
I tried both on GTK and the font looks exactly the same. Looking further into this, it is because GTK uses pango for font drawing. `m_size` isn't even used, it uses `m_wxfont`.

MaartenBent replied:
I've restructured `wxCairoFontData` constructor and added some comments, to address this and above comments.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -2749,6 +2762,31 @@ void wxCairoContext::Flush()
+    else
+    {
+        double dpi = (double)wxDisplay::GetStdPPIValue();
+#ifndef __WXMAC__
+        dpi *= GetContentScaleFactor();
+#endif
```
Again, this seems to indicate that there is something different with Cairo under Mac, but what?

vadz reviewed (commented):
Thanks a lot for all the work here!

It would be great if you could please answer my various questions below -- sorry if I'm missing something obvious, but some of the changes look puzzling (which doesn't mean that they're incorrect, of course). TIA!

csomor commented on samples/printing/printing.cpp:
```diff
@@ -142,79 +142,79 @@ void MyApp::Draw(wxDC&dc)
-    dc.DrawRectangle(0, 0, 230, 350);
-    dc.DrawLine(0, 0, 229, 349);
-    dc.DrawLine(229, 0, 0, 349);
+    dc.DrawRectangle(dc.FromDIP(0), dc.FromDIP(0), dc.FromDIP(230), dc.FromDIP(350));
```
would having some global version like `wxFromDIP` and `wxToDIP` with overloads taking wxPoints and wxRectangles (and returning the same type but converted) and a second parameter for wxDC, wxGraphicsContext etc. help to make things more readable ?

MaartenBent replied:
I used `FromDIP(0)` for consistency, removed it now.

I don't think adding more dip functions while make it better readable. Maybe a `DrawRectangle` that just accepts a rectangle to reduce the number of `FromDIP` (like `dc.DrawRectangle(dc.FromDIP(wxRect(0, 0, 230, 350)))`) but I wont do that in this PR.

csomor commented on src/generic/graphicc.cpp:
```diff
@@ -2237,7 +2250,7 @@ wxCairoContext::wxCairoContext( wxGraphicsRenderer* renderer, const wxMemoryDC&
 #endif
 
 #ifdef __WXMAC__
-    CGContextRef cgcontext = (CGContextRef)dc.GetWindow()->MacGetCGContextRef();
+    CGContextRef cgcontext = (CGContextRef)dc.GetSelectedBitmap().GetHBITMAP();
```
what is the reason for this change ? IIRC this bitmap-context does not know about its content scaling

MaartenBent replied:
Thanks, this works too (but saved images are the same, so maybe the bitmap-context does know about it).

csomor commented:
Thanks a lot @MaartenBent for accepting this challenge!

MaartenBent committed:
Fix Cairo DPI when no wxWindow is associated

Always multiply with the content scale factor, also on macOS.

MaartenBent commented on build/cmake/init.cmake:
```diff
@@ -574,6 +574,17 @@ if(wxUSE_GUI)
+            message(WARNING "Cairo not found, Cairo renderer won't be available")
+            wx_option_force_value(wxUSE_CAIRO OFF)
+        else()
+            #TODO only if build-in png?
```
This `else` is obsolete now and can be removed. It was for a different `FindCairo.cmake` that included all cairo dependent headers (lzo, libpng16, freetype2, harfbuzz, glib-2.0, pixman-1), which could cause problems if `imagpng.cpp` used the wrong png header.
The current one only adds `path/include/cairo` which is enough to build.

MaartenBent committed:
Simplify printing and drawing sample constructors

Don't use FromDIP for 0.
Include DIP in drawing sample variable and function names to make it clear these sizes are in DIP.
Show both logical and DIP coordinates in statusbar.

MaartenBent committed:
Remove wxWindow DPI from wxMemoryDC

wxMemoryDC is normally not associated with a wxWindow.

MaartenBent committed:
Fix printing at high DPI on GTK3 and when saving as postscript

MaartenBent committed:
CMake: Remove workaround for removing png headers from cairo

CAIRO_INCLUDE_DIRS does not include them with the current FindCairo.cmake.

MaartenBent committed:
Restructure Cairo font initialization

Add some comments explaining why things are done.

MaartenBent committed:
Use different method to get CGContextRef on Cairo renderer

MaartenBent commented on samples/drawing/drawing.cpp:
```diff
@@ -1999,8 +2000,8 @@ void MyCanvas::Draw(wxDC& pdc)
         dc.GetDeviceOrigin(&x0, &y0);
-        m_sizeX = dc.LogicalToDeviceX(dc.MaxX()) - x0 + 1;
-        m_sizeY = dc.LogicalToDeviceY(dc.MaxY()) - y0 + 1;
+        m_sizeX = dc.ToDIP(dc.LogicalToDeviceX(dc.MaxX()) - x0) + 1;
```
It's not only `wxSVGFileDC`, but `CreateWithDIPSize` also needs a size in DIP.
I updated the member and function names.

MaartenBent commented on samples/drawing/drawing.cpp:
```diff
@@ -2012,7 +2013,7 @@ void MyCanvas::OnMouseMove(wxMouseEvent &event)
         m_owner->PrepareDC(dc);
 
-        wxPoint pos = dc.DeviceToLogical(event.GetPosition());
+        wxPoint pos = dc.ToDIP(dc.DeviceToLogical(event.GetPosition()));
```
I updated it to show both logical coordinates and wxDC DIP coordinates (only when different).

MaartenBent commented on samples/printing/printing.cpp:
```diff
@@ -111,8 +111,8 @@ bool MyApp::OnInit(void)
-                                 wxPoint(0, 0), wxSize(400, 400));
-
+                                 wxDefaultPosition, wxDefaultSize);
+    frame->SetSize(frame->FromDIP(wxSize(400, 400)));
```
I updated the constructors in the drawing sample and printing sample so these unused arguments are not needed anymore.

MaartenBent commented on samples/printing/printing.cpp:
```diff
@@ -392,7 +392,7 @@ void MyFrame::OnPrintPreview(wxCommandEvent& WXUNUSED(event))
 
     wxPreviewFrame *frame =
-        new wxPreviewFrame(preview, this, "Demo Print Preview", wxPoint(100, 100), wxSize(600, 650));
+        new wxPreviewFrame(preview, this, "Demo Print Preview", wxPoint(100, 100), FromDIP(wxSize(600, 650)));
```
For a position it was not that important, so I left `FromDIP` out. Changed it to `wxDefaultPosition` now.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -1124,20 +1130,30 @@ wxCairoFontData::InitFontComponents(const wxString& facename,
+    : (font.GetFractionalPointSize() * dpi.y / 72);
 #else
+    // On macOS, font and screen both are 72 DPI, an macOS font size does
+    // not need a correction for retina displays. GTK does not use m_size.
```
Should we perhaps explain in this comment why do we still set `m_size` for it then? Or perhaps not set it, i.e. replace `#else` above with `#elif defined(__WXMAC__)`?

vadz commented on samples/drawing/drawing.cpp:
```diff
@@ -1896,15 +1887,18 @@ void MyCanvas::Draw(wxDC& pdc)
     if ( m_clip )
-        dc.SetClippingRegion(100, 100, 100, 100);
+    {
+        int clipSize = dc.FromDIP(100);
```
Sorry for nitpicking, but this is not a sizer, is it? I.e. it's used also as `x` and `y` arguments.

vadz reviewed (approved):
Looks perfect to me, thanks a lot!

I can just tweak the comment and the variable name mentioned below myself and merge already if you'd like, please let me know.

And thanks again!

MaartenBent commented:
I still need to do the interface files, so I can address the comments as well.

MaartenBent committed:
CMake: Mark Cairo and other module variables as advanced

MaartenBent committed:
Add documentation for wxDC and wxGraphicsContext FromDIP and ToDIP

Remove documentation of removed function GetDPIScaleFactor.

vadz commented:
Thanks a lot once again, AFAICS this should be merged and I'll do it soon unless @csomor has any comments/objections.

What would be a good way to briefly summarize CMake-related changes in this PR in the merge commit message?

MaartenBent commented:
> What would be a good way to briefly summarize CMake-related changes in this PR in the merge commit message?

Add support for finding Cairo when not using GTK, fix a warning when creating the inplace-config, and mark some variables as advanced.

MaartenBent committed:
Don't implement FromDIP and ToDIP in headers
